### ISSUE TITLE
Protos with a WKT as dep fail as they're on a _virtual_imports directory now

### ISSUE THREAD

**Issue by gertvdijk** [stackb/rules_proto#143]
IIUC Bazel 0.29.0+ started to put dependencies on WKTs for source protos on a `_virtual_imports/` path (with https://github.com/bazelbuild/bazel/commit/bcc9812e27d172d964dd728b70651fe0b972df0e) and fails for the approach this stack rules_proto takes in assembling the files into the 'staging area'.

A patch that I applied to make it work to indicate what I think should be changed:

```diff
diff --git compile.bzl compile.bzl
index 4f1d32c..1d0c96d 100644
--- compile.bzl
+++ compile.bzl
@@ -196,6 +196,11 @@ def _get_proto_filename(src):
         return "/".join(parts[2:])
     return src.short_path
 
+def _strip_virtual_import(path):
+      pos = path.find(_VIRTUAL_IMPORTS)
+      path = path[pos + len(_VIRTUAL_IMPORTS):]
+      return path.split("/", 1)[-1]
+
 def copy_proto(ctx, descriptor, src):
     """Copy a proto to the 'staging area'
 
@@ -207,12 +212,18 @@ def copy_proto(ctx, descriptor, src):
     Returns:
       <Generated File> for the copied .proto
     """
-    proto = ctx.actions.declare_file(_get_proto_filename(src), sibling = descriptor)
+
+    if is_in_virtual_imports(src):
+        proto_rpath = _strip_virtual_import(src.path)
+    else:
+        proto_rpath = _get_proto_filename(src)
+    proto_copy_path = "/".join([descriptor.dirname, proto_rpath])
+    proto = ctx.actions.declare_file(proto_rpath, sibling = descriptor)
     ctx.actions.run_shell(
         mnemonic = "CopyProto",
         inputs = [src],
         outputs = [proto],
-        command = "cp %s %s" % (src.path, proto.path),
+        command = "cp %s %s" % (src.path, proto_copy_path),
     )
     return proto
 
@@ -700,3 +711,21 @@ def invoke_transitive(proto_compile_rule, name_suffix, kwargs):
     )
 
     return rule_name
+
+# From https://github.com/grpc/grpc/blob/2e7d6b94eaf6b0e11add27606b4fe3d0b7216154/bazel/protobuf.bzl:
+
+_VIRTUAL_IMPORTS = "/_virtual_imports/"
+
+def is_in_virtual_imports(source_file, virtual_folder = _VIRTUAL_IMPORTS):
+    """Determines if source_file is virtual (is placed in _virtual_imports
+    subdirectory). The output of all proto_library targets which use
+    import_prefix  and/or strip_import_prefix arguments is placed under
+    _virtual_imports directory.
+    Args:
+        source_file: A proto file.
+        virtual_folder: The virtual folder name (is set to "_virtual_imports"
+            by default)
+    Returns:
+        True if source_file is located under _virtual_imports, False otherwise.
+    """
+    return not source_file.is_source and virtual_folder in source_file.path

base-commit: 6b334ece48828fb8e45052976d3516f808819ac7
-- 
2.25.0
```

Is this something acceptable for a PR?

**Comment by gmolau** [comment]
Any updates on this? I'm currently not able to get a grpc library with a WKT to build:

```
proto_library(
    name = "helloworld_proto",
    srcs = [
        "helloworld.proto",
    ],
    deps = [
        "@com_google_protobuf//:empty_proto",
    ],
)

python_grpc_library(
    name = "helloworld_py_grpc",
    deps = [
        ":helloworld_proto",
    ],
    has_services = True,
)
```

Building the grpc lib results in

```
ERROR: /private/var/tmp/_bazel_<user>/2e6771f21713012b2684c2011cd1e51c/external/com_google_protobuf/BUILD:309:2: error executing shell command: '/bin/bash -c bazel-out/host/bin/external/com_google_protobuf/protoc --descriptor_set_in=bazel-out/darwin-fastbuild/bin/external/com_google_protobuf/empty_proto-descriptor-set.proto.bin --python_out...' failed (Exit 1) bash failed: error executing command /bin/bash -c ... (remaining 1 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
Could not find file in descriptor database: src/google/protobuf/empty.proto: No such file or directory
```

**Comment by gertvdijk** [comment]
- I am still applying the patch. See https://github.com/TulipSolutions/tecl/tree/master/third_party/patches/com_github_stackb_rules_proto
- It seems that @pcj had started some refactoring that may fix this, see #142. (Haven't tried that branch.)

**Comment by pcj** [comment]
@gertvdijk I am trying out your patch and looks good, will create PR once I test it further.
